lib/perfdata/elasticsearchdatastreamwriter.cpp

```diff
@@ -0,0 +1,878 @@
+{
+	ObjectImpl<ElasticsearchDatastreamWriter>::OnConfigLoaded();
+
+	m_WorkQueue.SetName("ElasticsearchDatastreamWriter, " + GetName());
+
+	if (!GetEnableHa()) {
+		Log(LogDebug, "ElasticsearchDatastreamWriter")
+			<< "HA functionality disabled. Won't pause connection: " << GetName();
+
+		SetHAMode(HARunEverywhere);
+	} else {
+		SetHAMode(HARunOnce);
+	}
+}
+
+void ElasticsearchDatastreamWriter::StatsFunc(const Dictionary::Ptr& status, const Array::Ptr&)
+{
+	for (const ElasticsearchDatastreamWriter::Ptr& writer : ConfigType::GetObjectsByType<ElasticsearchDatastreamWriter>()) {
+		status->Set(
+			writer->GetName(),
+			new Dictionary({
+				{ "work_queue_items", writer->m_WorkQueue.GetTaskCount(60) / 60.0 },
+				{ "work_queue_item_rate", writer->m_WorkQueue.GetLength() },
+				{ "documents_sent", writer->m_DocumentsSent },
+				{ "documents_sent_error", writer->m_DocumentsFailed },
+				{ "checkresults_filtered_out", writer->m_ItemsFilteredOut.load() }
+			})
+		);
+	}
+}
+
+void ElasticsearchDatastreamWriter::Resume()
+{
+	ObjectImpl<ElasticsearchDatastreamWriter>::Resume();
+
+	Log(LogInformation, "ElasticsearchDatastreamWriter")
+		<< "'" << GetName() << "' resumed.";
+	m_Paused = false;
+	m_WorkQueue.SetExceptionCallback([](boost::exception_ptr exp) { ExceptionHandler(exp); });
+
+	if (GetManageIndexTemplate()) {
+		// Ensure index template exists/is updated as the first item on the work queue.
+		m_WorkQueue.Enqueue([this]() { ManageIndexTemplate(); });
+	}
+
+	/* Setup timer for periodically flushing m_DataBuffer */
+	m_FlushTimer = Timer::Create();
+	m_FlushTimer->SetInterval(GetFlushInterval());
+	m_FlushTimer->OnTimerExpired.connect([this](const Timer * const&) {
+		m_WorkQueue.Enqueue([this]() { Flush(); });
+	});
+	m_FlushTimer->Start();
+
+	/* Register for new metrics. */
+	m_HandleCheckResults = Checkable::OnNewCheckResult.connect(
+		[this](const Checkable::Ptr& checkable, const CheckResult::Ptr& cr, const MessageOrigin::Ptr&) {
+			CheckResultHandler(checkable, cr);
+		}
+	);
+}
+
+/* Pause is equivalent to Stop, but with HA capabilities to resume at runtime. */
+void ElasticsearchDatastreamWriter::Pause()
+{
+	m_HandleCheckResults.disconnect();
+	m_FlushTimer->Stop(true);
+	m_Paused = true;
+	m_WorkQueue.Enqueue([this]() { Flush(); });
+	m_WorkQueue.Join();
+
+	Log(LogInformation, "ElasticsearchDatastreamWriter")
+		<< "'" << GetName() << "' paused.";
+
+	ObjectImpl<ElasticsearchDatastreamWriter>::Pause();
+}
+
+void ElasticsearchDatastreamWriter::ManageIndexTemplate()
+{
+	AssertOnWorkQueue();
+
+	String templatePath = ICINGA_PKGDATADIR "/elasticsearch/index-template.json";
+	std::ifstream templateFile{ templatePath, std::ifstream::in };
+	if (!templateFile.is_open()) {
+		Log(LogCritical, "ElasticsearchDatastreamWriter")
+			<< "Could not open index template file: " << templatePath;
+		return;
+	}
+	std::ostringstream templateStream;
+	templateStream << templateFile.rdbuf();
+	String templateJson = templateStream.str();
```

Comment:
You can also read the entire file directly to the string variable like this and saves the extra copy from ostringstream.

```suggestion
	String templateJson((std::istreambuf_iterator<char>(templateFile)), std::istreambuf_iterator<char>());
```